frontend/src/main/java/com/google/tripmeout/frontend/servlets/PlaceIndividualServlet.java

```diff
@@ -0,0 +1,154 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class PlaceIndividualServlet extends HttpServlet {
+  private final PlaceVisitStorage placeStorage;
+  private final Gson gson;
+
+  private static final Pattern URI_NAME_PATTERN =
+      Pattern.compile(".*/trips/([^/]+)/placeVisits/([^/]+)");
+
+  private final GeoApiContext CONTEXT =
+      new GeoApiContext.Builder().apiKey("AIzaSyBbXCXC2uWv3baNmirLtqUYbFsFwCXqLV8").build();
+
+  public PlaceIndividualServlet(PlaceVisitStorage placeStorage, Gson gson) {
+    this.placeStorage = placeStorage;
+    this.gson = gson;
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    try {
+      Matcher matcher = ServletUtil.matchUriOrThrowError(request, URI_NAME_PATTERN);
+
+      String tripId = matcher.group(1);
+      String placeId = matcher.group(2);
+
+      Optional<PlaceVisitModel> place =
+          ServletUtil.extractFromRequestBody(request, gson, PlaceVisitModel.class);
+
+      if (!place.isPresent()) {
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      PlaceVisitModel.UserMark status = place.get().userMark();
+
+      PlaceDetailsRequest placeRequest = new PlaceDetailsRequest(CONTEXT).placeId(placeId).fields(
+          PlaceDetailsRequest.FieldMask.NAME, PlaceDetailsRequest.FieldMask.GEOMETRY_LOCATION);
+      PlaceDetails details = placeRequest.await();
+
+      PlaceVisitModel newPlace = PlaceVisitModel.builder()
+                                     .setTripId(tripId)
+                                     .setPlaceId(placeId)
+                                     .setName(details.name)
+                                     .setLatitude(details.geometry.location.lat)
+                                     .setLongitude(details.geometry.location.lng)
+                                     .build();
+
+      placeStorage.updateUserMarkOrAddPlaceVisit(newPlace, status);
+      response.setStatus(HttpServletResponse.SC_OK);
+
+    } catch (IllegalArgumentException e) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+
+    } catch (InvalidRequestException e) {
+      response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+
+    } catch (Exception e) {
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  @Override
+  public void doDelete(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    try {
+      Matcher matcher = ServletUtil.matchUriOrThrowError(request, URI_NAME_PATTERN);
+      String tripId = matcher.group(1);
+      String placeId = matcher.group(2);
+
+      placeStorage.removePlaceVisit(tripId, placeId);
+      response.setStatus(HttpServletResponse.SC_OK);
+
+    } catch (IllegalArgumentException e) {
```

Comment:
This might be fine for now (I'm actually not sure), but be careful: since `IllegalArgumentException` is a runtime exception, it may be hard to tell which methods this might originate from. If it's not strictly related to malformed/bad input, `SC_BAD_REQUEST` may be the wrong response code.

It seems like what we actually care about here is whether `ServletUtil.matchUriOrThrowError` threw an error, in which case we may want a custom checked exception that you can catch here. (Same thing applies to other servlet methods.)

Reply:
In a similar vein, the parse methods on `ServletUtil` should let the json parse exception escape instead of masking the error (and logging internally) and returning an empty optional. You can then catch that here and set the appropriate response code (SC_BAD_REQUEST).